### PR TITLE
Feature/issue 90 poisson logcdf

### PR DESCRIPTION
**Pull request by thelseraphim**
#### Submisison Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint`
- [ ] Declare copyright holder and open-source license: see below

#### Summary:
adds `poisson_log_cdf`, `poisson_log_cdf_log`, and `poisson_log_ccdf_log` and adds tests for correct values and partial derivatives with respect to the log rate parameter. this is computed in the silly way you'd expect, `poisson_log_cdf(n, t) = poisson_cdf(n, exp(t))` and the derivatives get a corresponding correction from the chain rule applied to the second argument.

resolves https://github.com/stan-dev/math/issues/90

#### How to Verify:
Besides running the six tests, it may be helpful to look at the adjacent files `poisson_cdf.hpp,` `poisson_cdf_log.hpp`, and `poisson_ccdf_log.hpp` of which these files are a minor variation. 

#### Side Effects:
None, adds three functions to the math library.

#### Documentation:
Will follow up with a PR adding these to the stan language and add doc in the manual at that time

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+      using std::exp;
+      using std::pow;
+
+      OperandsAndPartials<T_rate> operands_and_partials(log_lambda);
+
+      // Explicit return for extreme values
+      // The gradients are technically ill-defined, but treated as zero
+      for (size_t i = 0; i < stan::length(n); i++) {
+        if (value_of(n_vec[i]) < 0)
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: should we check for n being non-negative?

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+      for (size_t i = 0; i < size; i++) {
+        // Explicit results for extreme values
+        // The gradients are technically ill-defined, but treated as zero
+        if (value_of(n_vec[i]) == std::numeric_limits<int>::max())
+          continue;
+
+        const T_partials_return n_dbl = value_of(n_vec[i]);
+        const T_partials_return log_lambda_dbl = value_of(log_lambda_vec[i]);
+        const T_partials_return lambda_dbl = exp(log_lambda_dbl);
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: exp is expensive. Use a vector view for exp(log_lambda_dbl) instead.

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      }
+
+      for (size_t i = 0; i < size; i++) {
+        // Explicit results for extreme values
+        // The gradients are technically ill-defined, but treated as zero
+        if (value_of(n_vec[i]) == std::numeric_limits<int>::max())
+          continue;
+
+        const T_partials_return n_dbl = value_of(n_vec[i]);
+        const T_partials_return log_lambda_dbl = value_of(log_lambda_vec[i]);
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: Use a vector view for lambda instead -- there's no need for this temporary variable.

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        const T_partials_return lambda_dbl = exp(log_lambda_dbl);
+        const T_partials_return Pi = gamma_q(n_dbl+1, lambda_dbl);
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: Use a vector view for this.

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+        P *= Pi;
+
+        if (!is_constant_struct<T_rate>::value)
+          operands_and_partials.d_x1[i] -= exp(-lambda_dbl)
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: use a vector view for this.

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            * pow(lambda_dbl, n_dbl) / tgamma(n_dbl+1) / Pi;
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: and vector views for these terms.

File: stan/math/prim/scal/prob/poisson_log_cdf.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      }
+
+      if (!is_constant_struct<T_rate>::value) {
+        for (size_t i = 0; i < stan::length(log_lambda); ++i) {
+        const T_partials_return log_lambda_dbl = value_of(log_lambda_vec[i]);
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf.hpp]: fix spacing.

File: stan/math/prim/scal/prob/poisson_log_cdf_log.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
+#ifndef STAN_MATH_PRIM_SCAL_PROB_POISSON_LOG_CDF_LOG_HPP
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf_log.hpp]: Move this logic to poisson_log_lcdf.hpp. Change the logic of this file to call poisson_log_lcdf.

File: stan/math/prim/scal/prob/poisson_log_cdf_log.hpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+      using std::log;
+      using std::exp;
+
+      OperandsAndPartials<T_rate> operands_and_partials(log_lambda);
+
+      // Explicit return for extreme values
+      // The gradients are technically ill-defined, but treated as neg infinity
+      for (size_t i = 0; i < stan::length(n); i++) {
+        if (value_of(n_vec[i]) < 0)
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf_log.hpp]: should we check n for non-negative values outside and throw an exception?

File: stan/math/prim/scal/prob/poisson_log_cdf_log.hpp [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+      for (size_t i = 0; i < size; i++) {
+        // Explicit results for extreme values
+        // The gradients are technically ill-defined, but treated as zero
+        if (value_of(n_vec[i]) == std::numeric_limits<int>::max())
+          continue;
+
+        const T_partials_return n_dbl = value_of(n_vec[i]);
+        const T_partials_return log_lambda_dbl = value_of(log_lambda_vec[i]);
+        const T_partials_return lambda_dbl = exp(log_lambda_dbl);
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf_log.hpp]: please vectorize properly (see notes in other functions)

File: stan/math/prim/scal/prob/poisson_log_cdf_log.hpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        const T_partials_return log_Pi = log(gamma_q(n_dbl+1, lambda_dbl));
```
Comment (syclik) [inline stan/math/prim/scal/prob/poisson_log_cdf_log.hpp]: vectorize

File: test/unit/math/rev/scal/prob/poisson_log_ccdf_log_test.cpp [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+  std::vector<double> gradients1;
+  std::vector<double> gradients2;
+
+  val1.grad(x1, gradients1);
+  val2.grad(x2, gradients2);
+
+  EXPECT_FALSE(is_nan(gradients1[0]));
+  EXPECT_FALSE(is_nan(gradients2[0]));
+  EXPECT_NEAR(gradients1[0], gradients2[0] * rate_dbl, 0.01);
```
Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_ccdf_log_test.cpp]: change to `EXPECT_FLOAT_EQ`

Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_ccdf_log_test.cpp]: we shouldn't really need to use `EXPECT_NEAR` very often.

File: test/unit/math/rev/scal/prob/poisson_log_cdf_log_test.cpp [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+  std::vector<double> gradients1;
+  std::vector<double> gradients2;
+
+  val1.grad(x1, gradients1);
+  val2.grad(x2, gradients2);
+
+  EXPECT_FALSE(is_nan(gradients1[0]));
+  EXPECT_FALSE(is_nan(gradients2[0]));
+  EXPECT_NEAR(gradients1[0], gradients2[0] * rate_dbl, 0.01);
```
Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_cdf_log_test.cpp]: Change to `EXPECT_FLOAT_EQ`

File: test/unit/math/prim/scal/prob/poisson_log_ccdf_log_test.cpp [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+TEST(ProbPoisson, log_ccdf_log_values) {
+  int count = 2;
+  for (int i = 0; i < 6; ++i) {
+    double log_rate = 0.3;
+    for (int j = 0; j < 5; ++j) {
+      check_values(count, log_rate);
+    }
+  }
+}
```
Comment (syclik) [inline test/unit/math/prim/scal/prob/poisson_log_ccdf_log_test.cpp]: Add additional tests for boundary conditions.

File: test/unit/math/prim/scal/prob/poisson_log_cdf_log_test.cpp [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+TEST(ProbPoisson, log_cdf_log_values) {
+  int count = 2;
+  for (int i = 0; i < 6; ++i) {
+    double log_rate = 0.3;
+    for (int j = 0; j < 5; ++j) {
+      check_values(count, log_rate);
+    }
+  }
+}
```
Comment (syclik) [inline test/unit/math/prim/scal/prob/poisson_log_cdf_log_test.cpp]: Add additional tests for boundary conditions.

File: test/unit/math/prim/scal/prob/poisson_log_cdf_test.cpp [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+TEST(ProbPoisson, log_cdf_values) {
+  int count = 2;
+  for (int i = 0; i < 6; ++i) {
+    double log_rate = 0.3;
+    for (int j = 0; j < 5; ++j) {
+      check_values(count, log_rate);
+    }
+  }
+}
```
Comment (syclik) [inline test/unit/math/prim/scal/prob/poisson_log_cdf_test.cpp]: Add additional tests for boundary conditions.

File: test/unit/math/rev/scal/prob/poisson_log_ccdf_log_test.cpp [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+}
```
Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_ccdf_log_test.cpp]: Add additional tests for boundary conditions.

File: test/unit/math/rev/scal/prob/poisson_log_cdf_log_test.cpp [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+}
```
Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_cdf_log_test.cpp]: Add additional tests for boundary conditions.

File: test/unit/math/rev/scal/prob/poisson_log_cdf_test.cpp [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+  std::vector<double> gradients1;
+  std::vector<double> gradients2;
+
+  val1.grad(x1, gradients1);
+  val2.grad(x2, gradients2);
+
+  EXPECT_FALSE(is_nan(gradients1[0]));
+  EXPECT_FALSE(is_nan(gradients2[0]));
+  EXPECT_NEAR(gradients1[0], gradients2[0] * rate_dbl, 0.01);
```
Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_cdf_test.cpp]: Use `EXPECT_FLOAT_EQ`

File: test/unit/math/rev/scal/prob/poisson_log_cdf_test.cpp [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+}
```
Comment (syclik) [inline test/unit/math/rev/scal/prob/poisson_log_cdf_test.cpp]: Add additional tests for boundary conditions.

**Comment by thelseraphim** [comment]
@syclik thanks for the review. just a quick update on this: I created a separate issue (https://github.com/stan-dev/math/issues/475) and will modernize a bunch of functions in prim/scal/prob per your comments, starting with poisson_cdf, etc. Very few of the existing functions have any doc, clean vectorization, following their pattern was a mistake.

**Comment by syclik** [comment]
Any reason this PR was closed?

**Comment by thelseraphim** [comment]
oh just because I don't want to merge it, I'm going to rewrite poisson_cdf, etc, which are also not up to standards, and add these after, following their pattern.

**Comment by syclik** [comment]
Ok. Thanks.
